#482 #483 #491 Make word breaker testable and start writing tests.

danfickle opened the pull request:
We need a much more robust line breaker so any suggested unit tests will be incorporated.

danfickle committed:
#482 #483 #491 Make word breaker testable and start writing tests.

syjer commented:
I wonder if we could do a property based test in this case (using something like [junit quickcheck](https://github.com/pholser/junit-quickcheck/) ).

danfickle committed:
#482 #483 #491 Loads more tests for word breaker.

Plus minor behaviour change for word break method to avoid setting ends-on -soft-hyphen flag for soft hyphen at end of box.

danfickle committed:
#482 #483 #491 Add test to prove that too long words fall below floats.

This is to make sure infinite loop fixes do not break this functionality.

danfickle committed:
#429 #482 #483 #491 Safety valve on unbreakable word reattempts on new line.

With test to prove that we don't trigger safety valve with lots of unbreakable words. Safety valve should only be triggered if there is still a bug in our code.

danfickle committed:
#482 #483 #491 Basic fuzz testing of the line breaking algorithms.

This should ensure no infinite loop bugs creep in over time.

danfickle committed:
Merge branch 'open-dev-v1' into robust_breaker

danfickle commented:
This pr does the following:
+ Makes line breaker (by word) unit testable.
+ Adds several unit tests to exercise line breaker.
+ Adds end to end random fuzz testing for the line breaking. This includes all `white-space` values, both `break-word` and normal line breaking, with fonts which have a zero width soft hyphen as well as one that does not, in divs from 0px wide upwards. A random sampling of 5 character sequences (with the five characters chosen being special meaning in the line break algorithms) is used when the test is run each time, with the seed output to logs so any troublesome runs can be investigated. Hopefully, this random and regular testing will stop endless loop bugs creeping in again.
+ A safety valve for line breaking attempts when `break-word` is in effect. Triggering this safety check will deliberately result in a fatal exception which should be reported. Yes, this is a code smell, but one thought necessary after at least 3 endless loop bugs in my implementation of this complex functionality. We should be able to remove the safety check if the current fixed implementation proves robust.

danfickle commented:
> I wonder if we could do a property based test in this case (using something like [junit quickcheck](https://github.com/pholser/junit-quickcheck/) ).

I initially thought of parameter testing, but was underwhelmed by the syntax in Junit 4. After looking at the documentation of quickcheck, I probably should have used it rather than reinventing the wheel...
